File: patterns/structural/bridge.py
"""
Bridge Pattern - Python

Separates abstraction from implementation so both can vary independently.
Uses composition instead of inheritance.

Use cases:
- When you want to avoid permanent binding between abstraction and implementation
- When both abstractions and implementations should be extensible
- When changes in implementation should not affect clients
"""

from abc import ABC, abstractmethod


# Implementation interface
class Renderer(ABC):
    @abstractmethod
    def render_circle(self, radius):
        pass
    
    @abstractmethod
    def render_square(self, side):
        pass


# Concrete implementations
class VectorRenderer(Renderer):
    def render_circle(self, radius):
        return f"Drawing a circle of radius {radius} as vector graphics"
    
    def render_square(self, side):
        return f"Drawing a square of side {side} as vector graphics"


class RasterRenderer(Renderer):
    def render_circle(self, radius):
        return f"Drawing a circle of radius {radius} as pixels"
    
    def render_square(self, side):
        return f"Drawing a square of side {side} as pixels"


# Abstraction
class Shape(ABC):
    def __init__(self, renderer):
        self.renderer = renderer
    
    @abstractmethod
    def draw(self):
        pass


# Refined abstractions
class Circle(Shape):
    def __init__(self, renderer, radius):
        super().__init__(renderer)
        self.radius = radius
    
    def draw(self):
        return self.renderer.render_circle(self.radius)


class Square(Shape):
    def __init__(self, renderer, side):
        super().__init__(renderer)
        self.side = side
    
    def draw(self):
        return self.renderer.render_square(self.side)


# Example usage
if __name__ == "__main__":
    print("=== Bridge Pattern ===\n")
    
    vector_renderer = VectorRenderer()
    raster_renderer = RasterRenderer()
    
    vector_circle = Circle(vector_renderer, 5)
    raster_circle = Circle(raster_renderer, 5)
    
    print("Vector Circle:")
    print(vector_circle.draw())
    
    print("\nRaster Circle:")
    print(raster_circle.draw())
    
    vector_square = Square(vector_renderer, 10)
    raster_square = Square(raster_renderer, 10)
    
    print("\nVector Square:")
    print(vector_square.draw())
    
    print("\nRaster Square:")
    print(raster_square.draw())

File: patterns/structural/bridge.js
/**
 * Bridge Pattern - JavaScript
 * 
 * Separates abstraction from implementation so both can vary independently.
 * Uses composition instead of inheritance.
 * 
 * Use cases:
 * - When you want to avoid permanent binding between abstraction and implementation
 * - When both abstractions and implementations should be extensible
 * - When changes in implementation should not affect clients
 */

// Implementation interface
class Renderer {
    renderCircle(radius) {
        throw new Error('Method must be implemented');
    }
    
    renderSquare(side) {
        throw new Error('Method must be implemented');
    }
}

// Concrete implementations
class VectorRenderer extends Renderer {
    renderCircle(radius) {
        return `Drawing a circle of radius ${radius} as vector graphics`;
    }
    
    renderSquare(side) {
        return `Drawing a square of side ${side} as vector graphics`;
    }
}

class RasterRenderer extends Renderer {
    renderCircle(radius) {
        return `Drawing a circle of radius ${radius} as pixels`;
    }
    
    renderSquare(side) {
        return `Drawing a square of side ${side} as pixels`;
    }
}

// Abstraction
class Shape {
    constructor(renderer) {
        this.renderer = renderer;
    }
    
    draw() {
        throw new Error('Method must be implemented');
    }
}

// Refined abstractions
class Circle extends Shape {
    constructor(renderer, radius) {
        super(renderer);
        this.radius = radius;
    }
    
    draw() {
        return this.renderer.renderCircle(this.radius);
    }
}

class Square extends Shape {
    constructor(renderer, side) {
        super(renderer);
        this.side = side;
    }
    
    draw() {
        return this.renderer.renderSquare(this.side);
    }
}

// Example usage
console.log('=== Bridge Pattern ===\n');

const vectorRenderer = new VectorRenderer();
const rasterRenderer = new RasterRenderer();

const vectorCircle = new Circle(vectorRenderer, 5);
const rasterCircle = new Circle(rasterRenderer, 5);

console.log('Vector Circle:');
console.log(vectorCircle.draw());

console.log('\nRaster Circle:');
console.log(rasterCircle.draw());

const vectorSquare = new Square(vectorRenderer, 10);
const rasterSquare = new Square(rasterRenderer, 10);

console.log('\nVector Square:');
console.log(vectorSquare.draw());

console.log('\nRaster Square:');
console.log(rasterSquare.draw());

module.exports = { Renderer, VectorRenderer, RasterRenderer, Shape, Circle, Square };

